### PR TITLE
Redistribute hearing request cases

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves error that keeps priority push job from completing. See [this thread](https://dsva.slack.com/archives/CLRTL92TW/p1604318446004400) for info

### Description
Rather than creating an invalid transaction when we encounter a previously distributed case, just redistribute the case. Also alert sentry, just in case we want to know about it, to ensure the case should have been redistributed.

### Acceptance Criteria
- [x] When a duplicate distributed case exists, the priority push job does not fail

### Testing Plan
1. Create a previously distributed case
```ruby
# Some setup
JudgeTeam.for_judge(User.fourth).destroy!
RequestStore[:current_user] = User.system_user

# Make a hearing request appeal ready to distribute
task = DistributionTask.on_hold.where(appeal_id: Appeal.where(docket_type: Constants.AMA_DOCKETS.hearing)).last
task.appeal.hearings.update_all(disposition: "held")
task.children.first.completed!
task.appeal.reload.treee
                                             ┌────────────────────────────────────────────────────────────────────┐
Appeal 964 (hearing) ─────────────────────── │ ID   │ STATUS      │ ASGN_BY   │ ASGN_TO │ UPDATED_AT              │
└── RootTask                                 │ 2928 │ on_hold     │           │ Bva     │ 2020-10-30 14:18:58 UTC │
    ├── DistributionTask                     │ 2929 │ assigned    │           │ Bva     │ 2020-11-03 16:24:11 UTC │
    │   └── HearingTask                      │ 2930 │ completed   │ CSS_ID670 │ Bva     │ 2020-11-03 16:24:11 UTC │
    │       ├── ScheduleHearingTask          │ 2931 │ completed   │ CSS_ID671 │ Bva     │ 2020-10-30 14:18:58 UTC │
    │       └── AssignHearingDispositionTask │ 2932 │ in_progress │ CSS_ID672 │ Bva     │ 2020-10-30 14:18:58 UTC │
    └── TrackVeteranTask                     │ 2933 │ in_progress │           │ Vso     │ 2020-10-30 14:18:58 UTC │
                                             └────────────────────────────────────────────────────────────────────┘

# Make the appeal aod
task.appeal.claimant.person.update!(date_of_birth: 78.years.ago)
task.appeal.aod?
=> true

# Make a duplicate distributed case to cause this transaction to fail
dist = Distribution.create!(judge: User.third, priority_push: true)
DistributedCase.create!(case_id: task.appeal.uuid, distribution: dist, docket: "legacy", ready_at: Time.zone.now, priority: false, genpop: false, docket_index: 1, genpop_query: "any")
dist.update!(status: "completed")

# Run the job
PushPriorityAppealsToJudgesJob.new.perform_now
...
[2020-11-03 11:24:30 -0500]   DistributedCase Load (0.7ms)  SELECT  "distributed_cases".* FROM "distributed_cases" WHERE "distributed_cases"."case_id" = $1 LIMIT $2  [["case_id", "39cb729f-de4a-45f8-a9c5-8d58a62b8d80"], ["LIMIT", 1]]
Redistributing appeal 39cb729f-de4a-45f8-a9c5-8d58a62b8d80 to BVAAABSHIRE excluded from capture: DSN not set
[2020-11-03 11:24:30 -0500]   Distribution Load (1.2ms)  SELECT  "distributions".* FROM "distributions" WHERE "distributions"."id" = $1 LIMIT $2  [["id", 114], ["LIMIT", 1]]
[2020-11-03 11:24:30 -0500]   DistributedCase Update (0.6ms)  UPDATE "distributed_cases" SET "case_id" = $1, "updated_at" = $2 WHERE "distributed_cases"."id" = $3  [["case_id", "39cb729f-de4a-45f8-a9c5-8d58a62b8d80-redistributed-2020-11-03"], ["updated_at", "2020-11-03 16:24:30.546314"], ["id", 13]]
...
# Does not fail

dist.distributed_cases.first.case_id
=> "39cb729f-de4a-45f8-a9c5-8d58a62b8d80-redistributed-2020-11-03"

pp DistributedCase.where("case_id LIKE ?", "%#{task.appeal.uuid}%").pluck(:case_id)
["39cb729f-de4a-45f8-a9c5-8d58a62b8d80-redistributed-2020-11-03",
 "39cb729f-de4a-45f8-a9c5-8d58a62b8d80"]
```

### CODE REVIEW

File: app/models/vacols/case_docket.rb [hunk 1]
```diff
@@ -266,7 +266,7 @@ def self.age_of_oldest_priority_appeal
 
     fmtd_query = sanitize_sql_array([query, 1])
 
-    connection.exec_query(fmtd_query).to_hash.first["bfdloout"]
+    connection.exec_query(fmtd_query).to_hash.first&.fetch("bfdloout")
```
Comment (hschallhorn) [inline app/models/vacols/case_docket.rb:269]: fixes `NoMethodError (undefined method `[]' for nil:NilClass)` error

File: app/workflows/hearing_request_case_distributor.rb [hunk 2]
```diff
@@ -13,13 +13,18 @@ def initialize(appeals:, genpop:, distribution:, priority:)
...
-      Raven.capture_exception(error, extra: { uuid: appeal.uuid, judge: distribution.judge.css_id })
-      next
+    end
+  end
+
+  def redistribute_existing_distributed_case(appeal)
+    existing_case = DistributedCase.find_by(case_id: appeal.uuid)
+    if existing_case
+      Raven.capture_message("Redistributing appeal #{appeal.uuid} to #{distribution.judge.css_id}")
+      existing_case.redistribute!
```
Comment (hschallhorn) [inline app/workflows/hearing_request_case_distributor.rb:27]: Redistribute the case and alert sentry to ensure we keep track of these.

File: app/workflows/hearing_request_case_distributor.rb [hunk 3]
```diff
@@ -13,13 +13,18 @@ def initialize(appeals:, genpop:, distribution:, priority:)
   def call
     appeals_to_distribute.map do |appeal, genpop_value|
       Distribution.transaction do
+        redistribute_existing_distributed_case(appeal)
```
Comment (yoomlam) [inline app/workflows/hearing_request_case_distributor.rb:16]:
```suggestion
        rename_any_existing_distributed_case(appeal)
```
We should rename the `DistributedCase#redistribute` method -- it doesn't actually redistribute.

File: spec/models/dockets/hearing_request_docket_spec.rb [hunk 4]
```diff
@@ -231,7 +231,7 @@
         HearingRequestDocket.new.distribute_appeals(distribution, priority: false, limit: 10, genpop: "any")
       end
 
-      it "does not fail and distributes the legitimate tasks" do
+      it "does not fail, distributes the legitimate tasks, and redistributes the already distributed appeals" do
```
Comment (yoomlam) [inline spec/models/dockets/hearing_request_docket_spec.rb:234]:
```suggestion
      it "does not fail, renames conflicting already distributed appeals, and distributes the legitimate appeals" do
```

**Review by yoomlam — APPROVED** [review]
Approved with some suggestions.

**Comment by yoomlam** [comment]
For reference:
https://evilmartians.com/chronicles/the-silence-of-the-ruby-exceptions-a-rails-postgresql-database-transaction-thriller
https://github.com/ClosureTree/with_advisory_lock

**Pull request merged**